Add new gall guide. (#1249)
urbit/urbit.org: content/docs/userspace/gall-guide/types.md

+++
title = "Appendix: Types"
weight = 65
template = "doc.html"
+++

This document explains a few of the types commonly used in Gall agents. In
addition to these, the [Data Types](/docs/arvo/gall/data-types) section of the
Gall vane documentation is a useful reference. In particular, the
[`bitt`](/docs/arvo/gall/data-types#bitt),
[`boat`](/docs/arvo/gall/data-types#boat) and
[`bowl`](/docs/arvo/gall/data-types#bowl) entries, as well as the whole

> [`agent`](/docs/arvo/gall/data-types#agent) subsection.

Reviewer: Consider listing these in reverse order, so they're ordered most to least (likely to be) important.

## `vase`

Urbit needs to be able to send typed data between vanes and agents, as well as
over the network to other ships. While cell structures can easily be validated,

> atom auras cannot. To solve this problem, Urbit wraps such data in a vase before

Reviewer: I don't think this is the specific problem vases are intended to solve? Rather, it's just that vases let us encapsulate _dynamically_ typed data. In other words, this lets us send typed data around in contexts where we do not know the type ahead of time, where we can't have a _static_ type.

Also, we do not send vases over the network (to avoid having to coordinate on the type of `type` I believe), instead just sending around `cask`s. Internally/locally, yes, we do pass vases around.

sending it. Vases are used extensively - almost all data your agent will send
and received is wrapped in a vase.

A vase is just a cell with data in the tail and the type of the data in the
head. Its formal definition is:

```hoon
+$  vase  [p=type q=*]
```

Here's what it looks like if we bunt a vase in the dojo:

```
> *vase
[#t/* q=0]
```

There are two simple runes used to create and unpack vases. We'll look at each
of these next.

### Create a `vase`

The [zapgar](https://urbit.org/docs/hoon/reference/rune/zap#-zapgar) rune (`!>`)
takes a single argument of any noun, and wraps it in a vase. For example, in the
dojo:

```
> !>([1 2 3])
[#t/[@ud @ud @ud] q=[1 2 3]]

> !>('foo')
[#t/@t q=7.303.014]

> !>([[0xdead 0xb33f] 'foo'])
[#t/[[@ux @ux] @t] q=[[57.005 45.887] 7.303.014]]

> !>(foo='bar')
[#t/foo=@t q=7.496.034]
```

You would typically use `!>` as part of a [`cage`](#cage) when you're

> constructing a `card` like a poke or a `%fact` gift to be sent off.

Reviewer 2: ```suggestion
constructing a `card` like a poke or a `%fact` `gift` to be sent off.
```

### Extract data from `vase`

The [zapgal](https://urbit.org/docs/hoon/reference/rune/zap#-zapgal) rune (`!<`)
takes two arguments: A mold specifying the type to try and extract the data as,
and the vase to be extracted.

Let's look at an example in the dojo. First, let's create a vase of `[@t @ux @ud]`:

```
> =myvase !>(['foo' 0xabcd 123])
> myvase
[#t/[@t @ux @ud] q=[7.303.014 43.981 123]]
```

Next, let's try extracting our vase:

```
> !<  [@t @ux @ud]  myvase
['foo' 0xabcd 123]
```

Now let's try asking for a `@p` rather than `@t`:

```
> !<  [@p @ux @ud]  myvase
-need.@p
-have.@t
nest-fail
```

> As you can see, it will crash if the type does not exactly match. Note that

Reviewer 2: ```suggestion
As you can see, it will crash if the type does not nest. Note that
```

rather than using `!<`, you can also just clam the tail of the vase like:

```
> ((trel @t @ux @ud) +.myvase)
[p='foo' q=0xabcd r=123]
```

The only problem is that you can't tell if the auras were wrong:

```
> ((trel @p @ud @ux) +.myvase)
[p=~sibtel-tallyd q=43.981 r=0x7b]
```

You'd typically use `!<` on the data in `card`s that come in from other ships,
agents, etc.

## `mark`

The `mark` type is just a `@tas` like `%foo`, and specifies the Clay filetype of
some data. The `mark` corresponds to a mark file in the `/mar` directory, so a
`mark` of `%foo` corresponds to `/mar/foo/hoon`. Mark files are used for saving
data in Clay, validating data sent between agents or over the network, and
converting between different data types. For more information about mark files,
you can refer to the [Marks section of the Clay
documentation](/docs/arvo/clay/marks/marks).

## `cage`

A `cage` is a cell of a [`mark`](#mark) and a [`vase`](#vase), like `[%foo !>('bar')]`. The data in the vase should match the data type of the specified
mark.

Most data sent between agents, either locally or over the network, will be in a

> cage. When data comes in, Gall will validate the data in the vase with the given

Reviewer: As mentioned above, a `cask` goes over the network. The mark specified therein is usually used to "inflate" it back into a `cage`.

mark. Additionally, the data may be converted to a different mark if appropriate
conversion routines exist in the mark file. This is most commonly done when
talking to a web front-end via Eyre's channel system - outgoing data will be
automatically converted to a `%json` mark and incoming data will be converted
from `%json` to the mark your agent expects.

## `quip`

> `(quip a b)` is equivalent to `[(list a) b]`, it's just a more convenient way to

Reviewer: Feel like it might be nice to explicitly call this out as a type constructor instead of letting the context speak to that.

specify it. Most arms of an agent return a `(quip card _this)`, which is a list
of effects and a new state.

## `path`

The `path` type is formally defined as:

```hoon
+$  path  (list knot)
```

A knot is a `@ta` text atom (see the [Strings guide](/docs/hoon/guides/strings)
for details), so a `path` is just a list of text. Rather than having to write

> `[~.foo ~.bar ~.baz ~]` though, it has its own syntax which looks like

Reviewer 2: For whatever reason, the pretty printer puts out
```
> /foo/bar/baz
[%foo %bar %baz ~]
```
instead of `[~.foo ~.bar ~.baz ~]`. I guess this may just be a bug? Not sure if its worth mentioning here or not, but I know I've been confused by this before.

`/foo/bar/baz`.

> A `path` is just a general type used for several different purposes throughout

Reviewer 2: I think there could be a better one-liner than this. Maybe mention something like how paths are structured similarly to filepaths in Unix and play a role in giving data a location in a nested hierarchy.

Reviewer: > structured similarly to filepaths in Unix

But this is _also_ wrong! (^:

I agree the one-liner could be slightly less vague about things though. I think the general theme of its various uses is "identifier of some kind".

Arvo. Its elements have no inherent significance, it depends on the context. In
a Gall agent, a `path` is most commonly a subscription path - you might
subscribe for updates to `/foo/bar` on another agent, or another agent might
subscribe to `/baz` on your agent.

> A `path` might just be a series of fixed elements like `/foo/bar`, but some

Reviewer 2: ```suggestion
A `path` might just be a series of fixed `@ta` like `/foo/bar`, but some
```

> elements might also be variable and encode atom auras, or some other format. For

Reviewer: ```suggestion
A `path` might just be a series of fixed `@ta` like `/foo/bar`, but some
elements might also be variable and include encoded atoms, or some other datum. For
```

(Maybe also "encoded" -> "rendered"? Like, rendered to atom-syntax.)

Though I'll note it's generally discouraged to get too fancy with your encodings here. ie, you probably shouldn't `/my/path/(scot %uw (jam my-big-noun))` and then parse it back out later.

example, you might like to include a date in the path like
`/updates/~2021.10.31..07.24.27..db68`. Other agents might create the path by
doing something like:

```hoon
/update/(scot %da now.bowl)
```

Then, when you get a subscription request, you might do something like:

```hoon
?+    path  (on-watch:def path)
    [%updates @ ~]
  =/  date=@da  (slav %da i.t.path)

>   ...(rest of code)...

Reviewer: Might make the example slightly more minimal/focused:

```suggestion
?+    path  !!
    [%updates @ ~]
  =/  date=@da  (slav %da i.t.path)
  ...(rest of code)...
```

Though if the thrust here is to stay consistent with code shows in other parts of these docs, fair game.

```

See the [Encoding in text](/docs/hoon/guides/strings#encoding-in-text) and
[Decoding from text](/docs/hoon/guides/strings#decoding-from-text) sections of
the Strings guide for more information on dealing with atoms encoded in strings.

Aside from using function calls when constructing a `path` as demonstrated
above, you can also insert text you're previously stored with `=/` or what have
you, simply by enclosing them in brackets. For example, in the dojo:

```
> =const 'bar'
> `path`/foo/[const]/baz

> /foo/bar/baz

Reviewer: Notably using `@t` syntax (instead of `@ta`'s `~.bar`) here lets you construct "dirty" paths, that contain non-`@ta` symbols like spaces, which may prevent the rendered path from being parsed back into a noun in some places. Best to avoid, or at least be careful around.

```

## `wire`

The type of a wire is formally defined as:

```hoon
+$  wire  path
```

So, a `wire` is just a [`path`](#path), type-wise they're exactly the same. The
reason there's a separate `wire` type is just to differentiate their purpose. A
`wire` is a path for responses to requests an agent initiates. If you subscribe
to the `path` `/some/path` on another agent, you also specify `/some/wire`.
Then, when that agent sends out updates to subscribers of `/some/path`, your
agent receives them on `/some/wire`.

The `wire` is specified in the second argument of a `%pass` `card`. It's used
for anything you can `%pass`, such as `%poke`s, `%watch`es, and `%arvo` notes.
For example:

```hoon
[%pass /this/is/wire %agent [~zod %foobar] %watch /this/is/path]
::
[%pass /this/is/wire %agent [~zod %foobar] %poke %foo !>('hello')]
::
[%pass /this/is/wire %arvo %b %wait (add now.bowl ~m1)]
```

The `on-agent` and `on-arvo` arms of the agent core include a `wire` in their
respective sample. Responses from agents come in to the former, and responses

> from vanes come in to the latter.

Reviewer 2: May also consider briefly mentioning the role of `wire`s in call stacks and `duct`s, linking to `/docs/arvo/overview#types`

Then, when that agent sends out updates to subscribers of `/some/path`, your
agent receives them on `/some/wire`.

More formally, `wire`s are used by Arvo to represent an event cause, and
therefore return path, in a call stack called a
[`duct`](/docs/arvo/overview#duct). Inter-vane communications happen over
`duct`s as [`move`](/docs/arvo/overview#moves)s, and Gall converts the `card`s
produced by agents into such `move`s behind the scenes. A detailed understanding
of this system is not necessary to write Gall agents, but if you're interested
it's comprehensively documented in the [Arvo

> overview](/docs/arvo/overview#move).

Reviewer 2: ```suggestion
overview](/docs/arvo/overview) and [move trace tutorial](/docs/arvo/overview#move).
```

Reviewer 2: Just realized you removed the #move in the first link in https://github.com/urbit/urbit.org/pull/1249/commits/c206bc51471eb5ec1771bff715c45bdba97d6056 so I'll edit this
